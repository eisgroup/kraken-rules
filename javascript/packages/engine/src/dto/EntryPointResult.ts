/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from "./FieldEvaluationResult";
import { RuleEvaluationResults } from "./RuleEvaluationResults";
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult;

import { Reducer } from "declarative-js";
import flat = Reducer.flat;
import { ConditionEvaluationResult } from "./ConditionEvaluationResult";

/**
 * Contains evaluation results from specific entry point
 */
export class EntryPointResult {
    private readonly results: Record<string, FieldEvaluationResult>;
    readonly evaluationTimestamp: Date;
    constructor(results: Record<string, FieldEvaluationResult>) {
        this.results = results;
        this.getApplicableResults.bind(this);
        this.evaluationTimestamp = new Date();
    }
    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     */
    getAllRuleResults(): RuleEvaluationResult[] {
        return Object.values(this.results)
            .map(x => x.ruleResults)
            .reduce(flat, []);
    }
    getFieldResults(): Record<string, FieldEvaluationResult> {
        return this.results;
    }

    getApplicableResults(): RuleEvaluationResult[] {
        return this.getAllRuleResults()
            .filter(r => ConditionEvaluationResult.isApplicable(r.conditionEvaluationResult));
    }

    static empty(): EntryPointResult {
        return new EntryPointResult({});
    }
}
