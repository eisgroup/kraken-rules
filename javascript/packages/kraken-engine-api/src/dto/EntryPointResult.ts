/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { FieldEvaluationResult } from './FieldEvaluationResult'
import { RuleEvaluationResults } from './RuleEvaluationResults'
import RuleEvaluationResult = RuleEvaluationResults.RuleEvaluationResult

/**
 * Contains evaluation results from specific entry point
 */
export interface EntryPointResult {
    readonly evaluationTimestamp: Date
    readonly ruleTimeZoneId: string

    /**
     * Returns entry point result mapped by field name.
     */
    getFieldResults(): Record<string, FieldEvaluationResult>

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     */
    getAllRuleResults(): RuleEvaluationResult[]

    /**
     * Flattens and returns field results into an array of {@link RuleEvaluationResult}
     * filtered by rule condition evaluation status.
     */
    getApplicableResults(): RuleEvaluationResult[]
}
