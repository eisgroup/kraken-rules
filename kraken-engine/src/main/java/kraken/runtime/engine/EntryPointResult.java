package kraken.runtime.engine;

import kraken.annotations.API;
import kraken.runtime.engine.dto.FieldEvaluationResult;
import kraken.runtime.engine.dto.RuleEvaluationResult;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Contains evaluation results from specific entry point
 *
 * @author rimas
 * @since 1.0
 */
@API
public class EntryPointResult {

    /**
     * Field evaluation results, mapped to field id string
     */
    private Map<String, FieldEvaluationResult> fieldResults;

    private transient LocalDateTime evaluationTimeStamp;
    /**
     * Default constructor
     */
    public EntryPointResult() {
        fieldResults = new HashMap<>();
        evaluationTimeStamp = LocalDateTime.now();
    }

    public EntryPointResult(Map<String, FieldEvaluationResult> fieldResults, LocalDateTime evaluationTimeStamp) {
        this.fieldResults = new HashMap<>(fieldResults);
        this.evaluationTimeStamp = evaluationTimeStamp;
    }

    /**
     * Flatten contents into single list of {@link RuleEvaluationResult}s
     *
     * @return  list of all contained {@link RuleEvaluationResult}s
     */
    public List<RuleEvaluationResult> getAllRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .collect(Collectors.toList());
    }

    public List<RuleEvaluationResult> getApplicableRuleResults() {
        return fieldResults.entrySet().stream()
                .flatMap(e -> e.getValue().getRuleResults().stream())
                .filter(r -> r.getConditionEvaluationResult().isApplicable())
                .collect(Collectors.toList());
    }

    public LocalDateTime getEvaluationTimeStamp() {
        return evaluationTimeStamp;
    }

    public Map<String, FieldEvaluationResult> getFieldResults() {
        return fieldResults;
    }

}
