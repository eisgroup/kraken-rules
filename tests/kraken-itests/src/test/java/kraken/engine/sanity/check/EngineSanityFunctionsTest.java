/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.engine.sanity.check;

import kraken.runtime.EvaluationConfig;
import kraken.runtime.engine.EntryPointResult;
import kraken.testproduct.domain.*;
import kraken.testproduct.domain.extended.VehicleExtended;
import org.javamoney.moneta.Money;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import static io.github.jsonSnapshot.SnapshotMatcher.start;
import static io.github.jsonSnapshot.SnapshotMatcher.validateSnapshots;
import static kraken.testing.matchers.KrakenMatchers.*;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.is;
import static kraken.test.KrakenItestMatchers.matchesSnapshot;
import static org.hamcrest.MatcherAssert.assertThat;

public class EngineSanityFunctionsTest extends SanityEngineBaseTest {

    @BeforeClass
    public static void beforeAll() {
        start();
    }

    @AfterClass
    public static void afterAll() {
        validateSnapshots();
    }

    @Test
    public void shouldDefaultWithCountFunctionOutput() {
        final Policy policy = new Policy();
        policy.setPolicies(Arrays.asList("fourth", "sixth"));
        final HashMap<String, Object> context = new HashMap<>();
        context.put("numbers", List.of("first", "second", "third", "fifth"));
        EvaluationConfig evaluationConfig = new EvaluationConfig(context, "USD");
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Default-With-Count", evaluationConfig);
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(2));
        assertThat(policy.getPolicyNumber(), is("4"));
        assertThat(policy.getState(), is("2"));
    }

    @Test
    public void shouldDefaultWithSumFunctionOutput() {
        final Policy policy = new Policy();
        BillingInfo billingInfo = new BillingInfo();
        CreditCardInfo creditCardInfo = new CreditCardInfo();
        creditCardInfo.setCardCreditLimitAmount(Money.of(1555.35, "USD"));
        creditCardInfo.setCvv(123);
        billingInfo.setCreditCardInfo(creditCardInfo);
        policy.setBillingInfo(billingInfo);
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Default-With-Sum");
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(1));
        assertThat(policy.getPolicyNumber(), is("1678.35"));
    }

    @Test
    public void shouldDefaultWithAvgFunctionOutput() {
        final Policy policy = new Policy();
        Insured insured = new Insured();
        insured.setChildrenAges(Arrays.asList(5, 9, 18));
        policy.setInsured(insured);
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Default-With-Avg");
        assertThat(result, hasNoIgnoredRules());
        assertThat(result,  hasRuleResults(1));
        assertThat(policy.getPolicyNumber(), is("10.66666666666667"));
    }

    @Test
    public void shouldDefaultWithMaxFunctionOutput() {
        final Policy policy = new Policy();
        Vehicle vehicle = new Vehicle();
        policy.setRiskItems(List.of(vehicle));
        vehicle.setModelYear(2001);
        vehicle.setDeclaredAnnualMiles(1548458l);
        vehicle.setCostNew(BigDecimal.valueOf(12542.652));
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Default-With-Max");

        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(1));
        assertThat(vehicle.getNewValue(), is(BigDecimal.valueOf(1548458)));
    }

    @Test
    public void shouldDefaultWithMinFunctionOutput() {
        final Policy policy = new Policy();
        Insured insured = new Insured();
        insured.setChildrenAges(Arrays.asList(5, 9, 18));
        policy.setInsured(insured);
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Default-With-Min");
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(1));
        assertThat(policy.getPolicyNumber(), is("5"));
    }

    @Test
    public void shouldFlatElements() {
//        assert to have 7 items (3 of them serviceHistories)
        final Policy policy = new Policy();
        policy.setRiskItems(List.of(
                new VehicleExtended(List.of(LocalDate.MAX)),
                new VehicleExtended(List.of(LocalDate.MAX, LocalDate.MAX))
                )
        );
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Flat");
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(1));
        assertThat(result, hasValidationFailures(0));
    }
    @Test
    public void shouldFlatElementsAndFailAssertion() {
//        assert to have 7 items (2 of them serviceHistories)
        final Policy policy = new Policy();
        policy.setRiskItems(List.of(
                new VehicleExtended(List.of(LocalDate.MAX)),
                new VehicleExtended(List.of(LocalDate.MAX))
                )
        );
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Flat");
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(1));
        assertThat(result, hasValidationFailures(1));
    }

    @Test
    public void shouldDefaultWithSubstringOutput() {
        final Policy policy = new Policy();
        CreditCardInfo creditCardInfo = new CreditCardInfo();
        creditCardInfo.setCardNumber("378282246310005");
        creditCardInfo.setCardType("American Express");
        BillingInfo billingInfo = new BillingInfo();
        billingInfo.setCreditCardInfo(creditCardInfo);
        policy.setBillingInfo(billingInfo);
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Default-With-Substring");

        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(2));
        assertThat(policy.getPolicyNumber(), is("22463"));
        assertThat(policy.getState(), is(" Express"));
    }

    @Test
    public void shouldDefaultWithValueFromFunction() {
        final Policy policy = new Policy();
        TermDetails termDetails = new TermDetails();
        termDetails.setTermCd("cd125425464");
        policy.setTermDetails(termDetails);
        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Default-PolicyNumber");
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(1));
        assertThat(policy.getPolicyNumber(), is("125425464"));
    }

    @Test
    public void shouldExecuteFunctionAssertsOnPolicy() {
        final Policy policy = new Policy();
        final CreditCardInfo creditCardInfo = new CreditCardInfo();
        creditCardInfo.setCardCreditLimitAmount(Money.of(11, "USD"));
        policy.setBillingInfo(new BillingInfo("", creditCardInfo));

        final EntryPointResult result = engine.evaluate(policy, "FunctionCheck-Assert-PolicyNumber");
        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasNoValidationFailures());
        assertThat(result, hasRuleResults(29));
        assertThat(result, matchesSnapshot());
    }

    @Test
    public void shouldExecuteKelFunction() {
        COLLCoverage coverage1 = new COLLCoverage(new BigDecimal(50));
        COLLCoverage coverage2 = new COLLCoverage(new BigDecimal(25));
        Vehicle vehicle1 = new Vehicle();
        vehicle1.setCollCoverages(List.of(coverage1, coverage2));
        COLLCoverage coverage3 = new COLLCoverage(new BigDecimal(125));
        Vehicle vehicle2 = new Vehicle();
        vehicle2.setCollCoverages(List.of(coverage3));
        Policy policy = new Policy();
        policy.setTransactionDetails(new TransactionDetails());
        policy.setRiskItems(List.of(vehicle1, vehicle2));

        EntryPointResult result = engine.evaluate(policy, "Expressions_kel_functions");

        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasNoValidationFailures());
        assertThat(result, hasRuleResults(2));
        assertThat(policy.getTransactionDetails().getTotalLimit(), equalTo(new BigDecimal(200)));
    }

    @Test
    public void shouldExecuteFromMoneyFunctionAndPassAssertions() {
        Policy policy = new Policy();
        CreditCardInfo creditCardInfo = new CreditCardInfo();
        creditCardInfo.setCardCreditLimitAmount(Money.of(100, "USD"));
        policy.setBillingInfo(new BillingInfo("ACC", creditCardInfo));

        EntryPointResult result = engine.evaluate(policy, "FunctionCheck-FromMoney");

        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasNoValidationFailures());
        assertThat(result, hasRuleResults(1));
        assertThat(result, matchesSnapshot());
    }

    @Test
    public void shouldExecuteFromMoneyFunctionAndFailAssertions() {
        Policy policy = new Policy();
        CreditCardInfo creditCardInfo = new CreditCardInfo();
        creditCardInfo.setCardCreditLimitAmount(Money.of(99, "USD"));
        policy.setBillingInfo(new BillingInfo("ACC", creditCardInfo));

        EntryPointResult result = engine.evaluate(policy, "FunctionCheck-FromMoney");

        assertThat(result, hasNoIgnoredRules());
        assertThat(result, hasRuleResults(1));
        assertThat(result, hasValidationFailures(1));
        assertThat(result, matchesSnapshot());
    }

}
