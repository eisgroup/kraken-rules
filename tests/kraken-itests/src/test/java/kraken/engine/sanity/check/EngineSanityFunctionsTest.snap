kraken.engine.sanity.check.EngineSanityFunctionsTest.shouldExecuteFromMoneyFunctionAndFailAssertions=[
  {
    "CreditCardInfo:-1:cardCreditLimitAmount": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "CreditCardInfo",
        "fieldName": "cardCreditLimitAmount",
        "fieldPath": "cardCreditLimitAmount"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "f-fromMoney-explicit-usage",
            "context": "CreditCardInfo",
            "targetPath": "cardCreditLimitAmount",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": false,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]


kraken.engine.sanity.check.EngineSanityFunctionsTest.shouldExecuteFromMoneyFunctionAndPassAssertions=[
  {
    "CreditCardInfo:-1:cardCreditLimitAmount": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "CreditCardInfo",
        "fieldName": "cardCreditLimitAmount",
        "fieldPath": "cardCreditLimitAmount"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "f-fromMoney-explicit-usage",
            "context": "CreditCardInfo",
            "targetPath": "cardCreditLimitAmount",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]


kraken.engine.sanity.check.EngineSanityFunctionsTest.shouldExecuteFunctionAssertsOnPolicy=[
  {
    "Policy:-1:policyNumber": {
      "contextFieldInfo": {
        "contextId": "-1",
        "contextName": "Policy",
        "fieldName": "policyNumber",
        "fieldPath": "policyNumber"
      },
      "ruleResults": [
        {
          "ruleInfo": {
            "ruleName": "f-symmetricDifference-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-distinct-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-upper-case-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-includes-in-array-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-format-date-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-intersection-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-trim-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-pad-left-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-all-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-union-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-difference-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-concat-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-pad-right-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-money-to-string-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-join-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-includes-in-string-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-starts-with-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-ends-with-between-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-lower-case-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-get-month-date-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-get-day-date-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-isBlank-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-get-year-date-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-date-months-between-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-length-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-number-to-string-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-any-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-isEmpty-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        },
        {
          "ruleInfo": {
            "ruleName": "f-numberSequence-assert-policyNumber",
            "context": "Policy",
            "targetPath": "policyNumber",
            "payloadType": "ASSERTION"
          },
          "payloadResult": {
            "success": true,
            "templateVariables": [],
            "rawTemplateVariables": [],
            "validationSeverity": "critical"
          },
          "conditionEvaluationResult": {
            "conditionEvaluation": "APPLICABLE"
          },
          "overrideInfo": {
            "isOverridable": false
          },
          "ruleEvaluationStatus": "APPLIED"
        }
      ]
    }
  }
]